File: sdk/golang/rooch-sdk/bcs/serializer.go
package bcs

import (
	"fmt"
	"regexp"
	"strings"

	"github.com/rooch-network/rooch/sdk/golang/rooch-sdk/address"
	"github.com/rooch-network/rooch/sdk/golang/rooch-sdk/utils"
)

var (
	vectorRegex = regexp.MustCompile(`^vector<(.+)>$`)
	structRegex = regexp.MustCompile(`^([^:]+)::([^:]+)::([^<]+)(<(.+)>)?`)
)

type Serializer struct{}

func (s *Serializer) StructTagToObjectID(input StructTag) string {
	return "0x" + utils.ToHex(utils.Sha3256(s.StructTagToCanonicalString(input)))
}

func (s *Serializer) StructTagToCanonicalString(input StructTag) string {
	result := fmt.Sprintf("%s::%s::%s", address.CanonicalRoochAddress(input.Address), input.Module, input.Name)

	if len(input.TypeParams) > 0 {
		typeParams := make([]string, len(input.TypeParams))
		for i, param := range input.TypeParams {
			typeParams[i] = s.TypeTagToString(param)
		}
		result += fmt.Sprintf("<%s>", strings.Join(typeParams, ","))
	}

	return result
}

func (s *Serializer) TypeTagToString(input TypeTag) string {
	if str, ok := input.(string); ok {
		return str
	}

	if vector, ok := input.(map[string]TypeTag); ok {
		if val, exists := vector["Vector"]; exists {
			return fmt.Sprintf("vector<%s>", s.TypeTagToString(val))
		}
	}

	if structTag, ok := input.(map[string]StructTag); ok {
		if val, exists := structTag["Struct"]; exists {
			return s.StructTagToCanonicalString(val)
		}
	}

	panic("Invalid TypeTag")
}

func (s *Serializer) TypeTagParseFromStr(str string, normalizeAddress bool) BcsTypeTag {
	switch str {
	case "address":
		return BcsTypeTag{Address: &struct{}{}}
	case "bool":
		return BcsTypeTag{Bool: &struct{}{}}
	case "u8":
		return BcsTypeTag{U8: &struct{}{}}
	case "u16":
		return BcsTypeTag{U16: &struct{}{}}
	case "u32":
		return BcsTypeTag{U32: &struct{}{}}
	case "u64":
		return BcsTypeTag{U64: &struct{}{}}
	case "u128":
		return BcsTypeTag{U128: &struct{}{}}
	case "u256":
		return BcsTypeTag{U256: &struct{}{}}
	case "signer":
		return BcsTypeTag{Signer: &struct{}{}}
	}

	if matches := vectorRegex.FindStringSubmatch(str); matches != nil {
		return BcsTypeTag{
			Vector: s.TypeTagParseFromStr(matches[1], normalizeAddress),
		}
	}

	if matches := structRegex.FindStringSubmatch(str); matches != nil {
		addr := matches[1]
		if normalizeAddress {
			addr = address.NormalizeRoochAddress(addr)
		}
		
		var typeParams []BcsTypeTag
		if matches[5] != "" {
			typeParams = s.ParseStructTypeArgs(matches[5], normalizeAddress)
		}

		return BcsTypeTag{
			Struct: &StructTag{
				Address:    addr,
				Module:     matches[2],
				Name:       matches[3],
				TypeParams: typeParams,
			},
		}
	}

	panic(fmt.Sprintf("Encountered unexpected token when parsing type args for %s", str))
}

func (s *Serializer) ParseStructTypeArgs(str string, normalizeAddress bool) []BcsTypeTag {
	tokens := SplitGenericParameters(str)
	result := make([]BcsTypeTag, len(tokens))
	for i, token := range tokens {
		result[i] = s.TypeTagParseFromStr(token, normalizeAddress)
	}
	return result
}

func (s *Serializer) TagToString(tag BcsTypeTag) string {
	switch {
	case tag.Bool != nil:
		return "bool"
	case tag.U8 != nil:
		return "u8"
	case tag.U16 != nil:
		return "u16"
	case tag.U32 != nil:
		return "u32"
	case tag.U64 != nil:
		return "u64"
	case tag.U128 != nil:
		return "u128"
	case tag.U256 != nil:
		return "u256"
	case tag.Address != nil:
		return "address"
	case tag.Signer != nil:
		return "signer"
	case tag.Vector != nil:
		return fmt.Sprintf("vector<%s>", s.TagToString(*tag.Vector))
	case tag.Struct != nil:
		struct_ := tag.Struct
		typeParams := make([]string, len(struct_.TypeParams))
		for i, param := range struct_.TypeParams {
			typeParams[i] = s.TagToString(param)
		}
		typeParamsStr := ""
		if len(typeParams) > 0 {
			typeParamsStr = fmt.Sprintf("<%s>", strings.Join(typeParams, ", "))
		}
		return fmt.Sprintf("%s::%s::%s%s", struct_.Address, struct_.Module, struct_.Name, typeParamsStr)
	}
	panic("Invalid TypeTag")
} 

File: sdk/typescript/rooch-sdk/src/bcs/serializer.ts
// Copyright (c) RoochNetwork
// SPDX-License-Identifier: Apache-2.0

import { splitGenericParameters } from '@mysten/bcs'

import { concatBytes, isBytes, sha3_256, stringToBytes, toHEX } from '../utils/index.js'
import { canonicalRoochAddress, normalizeRoochAddress, RoochAddress } from '../address/index.js'

import { StructTag, TypeTag, BcsTypeTag } from './types.js'
import { address } from '../types/rooch.js'
import { Bytes } from '../types/bytes.js'

const VECTOR_REGEX = /^vector<(.+)>$/
const STRUCT_REGEX = /^([^:]+)::([^:]+)::([^<]+)(<(.+)>)?/

export class Serializer {
  static structTagToObjectID(input: StructTag | string): string {
    return `0x${toHEX(sha3_256(typeof input === 'string' ? input : Serializer.structTagToCanonicalString(input)))}`
  }

  static accountNamedObjectID(address: address, structTag: StructTag | string) {
    let addressBytes: Bytes
    if (typeof address === 'string') {
      const normalizeAddress = normalizeRoochAddress(address)
      addressBytes = new RoochAddress(normalizeAddress).toBytes()
    } else if (isBytes(address)) {
      addressBytes = address
    } else {
      addressBytes = address.toBytes()
    }

    const tagBytes = stringToBytes(
      'utf8',
      typeof structTag === 'string' ? structTag : Serializer.structTagToCanonicalString(structTag),
    )

    return `0x${toHEX(sha3_256(concatBytes(addressBytes, tagBytes)))}`
  }

  static structTagToCanonicalString(input: StructTag): string {
    let result = `${canonicalRoochAddress(input.address)}::${input.module}::${input.name}`

    if (input.typeParams && input.typeParams.length > 0) {
      const typeParams = input.typeParams.map(Serializer.typeTagToString).join(',')
      result += `<${typeParams}>`
    }

    return result
  }

  static typeTagToString(input: TypeTag): string {
    if (typeof input === 'string') {
      return input
    }

    if ('Vector' in input) {
      return `vector<${Serializer.typeTagToString(input.Vector)}>`
    }

    if ('Struct' in input) {
      return Serializer.structTagToCanonicalString(input.Struct)
    }

    throw new Error('Invalid TypeTag')
  }

  static typeTagParseFromStr(str: string, normalizeAddress = false): BcsTypeTag {
    if (str === 'address') {
      return { address: null }
    } else if (str === 'bool') {
      return { bool: null }
    } else if (str === 'u8') {
      return { u8: null }
    } else if (str === 'u16') {
      return { u16: null }
    } else if (str === 'u32') {
      return { u32: null }
    } else if (str === 'u64') {
      return { u64: null }
    } else if (str === 'u128') {
      return { u128: null }
    } else if (str === 'u256') {
      return { u256: null }
    } else if (str === 'signer') {
      return { signer: null }
    }

    const vectorMatch = str.match(VECTOR_REGEX)
    if (vectorMatch) {
      return {
        vector: Serializer.typeTagParseFromStr(vectorMatch[1], normalizeAddress),
      }
    }

    const structMatch = str.match(STRUCT_REGEX)
    if (structMatch) {
      const address = normalizeAddress ? normalizeRoochAddress(structMatch[1]) : structMatch[1]
      return {
        struct: {
          address,
          module: structMatch[2],
          name: structMatch[3],
          typeParams:
            structMatch[5] === undefined
              ? []
              : Serializer.parseStructTypeArgs(structMatch[5], normalizeAddress),
        },
      }
    }

    throw new Error(`Encountered unexpected token when parsing type args for ${str}`)
  }

  static parseStructTypeArgs(str: string, normalizeAddress = false): BcsTypeTag[] {
    return splitGenericParameters(str).map((tok) =>
      Serializer.typeTagParseFromStr(tok, normalizeAddress),
    )
  }

  static tagToString(tag: BcsTypeTag): string {
    if ('bool' in tag) {
      return 'bool'
    }
    if ('u8' in tag) {
      return 'u8'
    }
    if ('u16' in tag) {
      return 'u16'
    }
    if ('u32' in tag) {
      return 'u32'
    }
    if ('u64' in tag) {
      return 'u64'
    }
    if ('u128' in tag) {
      return 'u128'
    }
    if ('u256' in tag) {
      return 'u256'
    }
    if ('address' in tag) {
      return 'address'
    }
    if ('signer' in tag) {
      return 'signer'
    }
    if ('vector' in tag) {
      return `vector<${Serializer.tagToString(tag.vector)}>`
    }
    if ('struct' in tag) {
      const struct = tag.struct
      const typeParams = struct.typeParams.map(Serializer.tagToString).join(', ')
      return `${struct.address}::${struct.module}::${struct.name}${
        typeParams ? `<${typeParams}>` : ''
      }`
    }
    throw new Error('Invalid TypeTag')
  }
}
